File: 109_Convert_Sorted_List_to_Binary_Search_Tree/109.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sortedListToBST(self, head):
        """
        :type head: ListNode
        :rtype: TreeNode
        """
        
        if head is None: 
            return None
        if head.next is None:
            return TreeNode(head.val)
        
        def FindMid(self, head):
            slow = fast = head
            prev = None
            while fast and fast.next:
                prev = slow
                slow = slow.next
                fast = fast.next.next
            prev.next = None
            return slow
            
        mid = FindMid(self, head)
     
        root = TreeNode(mid.val)
        root.left = self.sortedListToBST(head)
        root.right = self.sortedListToBST(mid.next)

        return root

File: 109_Convert_Sorted_List_to_Binary_Search_Tree/109.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    
    TreeNode* sortedListToBST(ListNode* head) {
        if (!head) return NULL;
        if (!head->next) return new TreeNode(head->val); 
        
        // Find Mid
        ListNode* slow = head, *fast = head;
        ListNode* prev = head;
        while (fast and fast->next)
        {
            prev = slow;
            slow = slow->next;
            fast = fast->next->next;
            
        }
        prev->next = NULL;
        
        TreeNode *root =  new TreeNode(slow->val); 
        root->right = sortedListToBST (slow->next);
        root->left  = sortedListToBST (head);
        
        return root;
    }
 
};